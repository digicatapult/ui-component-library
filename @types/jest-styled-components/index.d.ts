declare module 'jest-styled-components/serializer'
