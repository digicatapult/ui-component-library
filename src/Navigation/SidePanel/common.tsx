import styled from 'styled-components'
import { SidePanelProps } from './index.js'

export const Heading = styled('div')`
  color: #4a4a4a;
  font-family: Roboto;
  margin-bottom: 100px;
  max-width: 80%;
  font-size: 30px;
  line-height: 45px;
`

export const Button = styled.div<SidePanelProps>`
  position: absolute;
  top: 10%;
  right: 0px;
  cursor: pointer;
  transform: scaleX(${({ isOpen }) => (isOpen ? '-1' : '1')}) translateX(100%);
  transform-origin: 100%;
  transition: transform 0.7s ease-in-out;
`

export const ItemWrapper = styled('div')`
  transition: all 0.3s ease-out;
  &:hover {
    opacity: 0.7;
  }
`

export const DefaultPanel = styled.div<SidePanelProps>`
  position: ${({ variant }) => (variant === 'hyproof' ? 'fixed' : 'absolute')};
  box-sizing: border-box;
  width: ${({ width }) => width || '300px'};
  height: 100lvh;
  left: 0px;
  top: 0px;

  display: flex;
  flex-direction: column;
  gap: 0.2em;

  padding: 5px 25px;
  transform: translateX(
    ${({ isOpen, width }) => (isOpen ? '0px' : `-${width || '300px'}`)}
  );
  transition: transform 0.7s ease-in-out;

  box-shadow: 0px 4px 5px rgba(0, 0, 0, 0.1);
`

export const HiiVariantPanel = styled(DefaultPanel)<SidePanelProps>`
  box-shadow: none;
  background: rgba(255, 255, 255, 90%);
  backdrop-filter: blur(5px);
`
