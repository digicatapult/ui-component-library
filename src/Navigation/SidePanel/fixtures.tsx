export const basic = [
  {
    name: 'some-basic-name-1',
    title: 'title of the card',
    subtitle: 'and some subtitle',
  },
  {
    name: 'some-basic-name-2',
    title: 'no subtitle',
  },
  {
    name: 'some-basic-name-2',
    title: 'no subtitle',
  },
]

export const hyproof = [
  {
    id: 'heidi',
    title: 'Heidi Heidi',
    subtitle: 'The Hydrogen Producer',
    background: '#9EDCFA',
    active: true,
  },
  {
    id: 'emma',
    title: 'Emma Emma',
    subtitle: 'The Energy Producer',
    background: '#AAED93',
  },
  {
    id: 'conor',
    title: 'Conor Conor',
    subtitle: 'The Hydrogen Consumer',
    background: '#FDB6D4',
  },
  {
    id: 'reginald',
    title: 'Reginald Reginald',
    subtitle: 'The Regulator',
    background: '#FCF281',
  },
]
