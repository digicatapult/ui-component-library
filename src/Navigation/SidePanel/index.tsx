import React from 'react'

import { Heading, Button, Panel, ItemWrapper } from './common.js'
import ListCard, { ListCardProps } from '../../ListCard/index.js'
import { SideArrowCloseIcon, SideArrowOpenIcon } from '../../index.js'
import Avatar from '../../UserIcon/index.js'

export interface SidePanelProps {
  heading?: string
  title?: string
  variant?: 'default' | 'hyproof'
  width?: string
  isOpen?: boolean
  callback?: (data: { [k: string]: string | number | boolean }) => void
  update?: (name: string, persona: Persona, e?: MouseEvent) => void
}

export interface SidePanelItemProps extends SidePanelProps {
  name: string
  title: string
  background?: string
  color?: string
  subtitle?: string
}

interface IItem {
  Item: typeof Item
}
type Persona = Omit<ListCardProps, 'onClick'>
type ISidePanel = React.FC<React.PropsWithChildren<SidePanelProps>>

const Item: React.FC<React.PropsWithChildren<SidePanelItemProps>> = ({
  update = () => {},
  variant = 'default',
  ...props
}) => (
  <ItemWrapper>
    <ListCard
      variant={variant}
      Icon={() => {
        if (variant === 'hyproof')
          return (
            <Avatar
              bgColor={props.background}
              color={props.color}
              fullName={props.name}
              outlineColor="white"
              size="70px"
            />
          )
      }}
      onClick={(name: any) => {
        update(name, props)
      }}
      title={props.title}
      subtitle={props.subtitle}
    />
  </ItemWrapper>
)

const SidePanel: ISidePanel & IItem = ({
  children,
  update,
  callback,
  heading,
  ...props
}) => {
  const [show, setShow] = React.useState(false)

  const style = { width: props.width, isOpen: show } as React.CSSProperties
  return (
    <Panel onAnimationEnd={() => setShow(true)} style={style}>
      <Button
        onClick={() => {
          setShow(!show)
          if (callback) return callback({ ...props, isOpen: !show })
        }}
        style={style}
      >
        {show ? <SideArrowOpenIcon /> : <SideArrowCloseIcon />}
      </Button>
      <Heading>{`${heading} (${props.title})`}</Heading>
      {children}
    </Panel>
  )
}

SidePanel.Item = Item

export default SidePanel
