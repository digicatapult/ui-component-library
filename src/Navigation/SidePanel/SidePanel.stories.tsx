import React from 'react'
import { action } from '@storybook/addon-actions'

import SidePanel, { SidePanelProps } from './index.js'
import { hyproof, basic } from './fixtures.js'

const fixtures = { hyproof, default: basic }

export default {
  title: 'Components/Navigation/SidePanel',
  component: SidePanel,
  argTypes: {
    variant: {
      options: ['default', 'hyproof'],
      control: { type: 'radio' },
    },
    update: { action: 'submit' },
    heading: {
      control: 'text',
    },
    color: {
      control: {
        type: 'select',
      },
      options: ['#a7a7a7', '#fff', '#00ff00', '#33e58c', 'red'],
    },
  },
}

const DefaultStoryTemplate = (args: SidePanelProps) => {
  const [current, setCurrent] = React.useState({})

  return (
    <SidePanel {...args} {...current} callback={action('open')}>
      {fixtures[args.variant || 'default'].map((item) => (
        <SidePanel.Item
          {...args}
          update={(name, persona) => {
            setCurrent({ current: name, ...persona })
          }}
          key={item.name}
          {...item}
        />
      ))}
    </SidePanel>
  )
}

export const Default = DefaultStoryTemplate.bind({})
Default.args = {
  variant: 'default',
  heading: 'Basic Side Panel',
  width: 300,
}

export const HyProof = DefaultStoryTemplate.bind({})
HyProof.args = {
  variant: 'hyproof',
  heading: 'HyProof Certificate Viewing',
  width: 400,
}