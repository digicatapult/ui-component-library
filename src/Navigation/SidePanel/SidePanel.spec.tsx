import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import SidePanel from './index.js'

describe('SidePanel', () => {
  test('default', () => {
    const instance = renderer.create(<SidePanel title="title" />)
    const tree = instance.toJSON()

    expect(tree).toMatchSnapshot()
  })
})
