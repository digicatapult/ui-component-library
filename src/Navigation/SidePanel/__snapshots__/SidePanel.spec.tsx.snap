// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SidePanel default 1`] = `
.c2 {
  color: #4a4a4a;
  font-family: Roboto;
  margin-bottom: 100px;
  max-width: 80%;
  font-size: 30px;
  line-height: 45px;
}

.c1 {
  position: absolute;
  top: 10%;
  right: 0px;
  cursor: pointer;
  transform: scaleX(1) translateX(100%);
  transform-origin: 100%;
  transition: transform 0.7s ease-in-out;
}

.c3 {
  transition: all 0.3s ease-out;
}

.c3:hover {
  opacity: 0.7;
}

.c0 {
  position: absolute;
  box-sizing: border-box;
  width: 300px;
  height: 100lvh;
  left: 0px;
  top: 0px;
  display: flex;
  flex-direction: column;
  gap: 0.2em;
  padding: 5px 25px;
  transform: translateX(     -300px   );
  transition: transform 0.7s ease-in-out;
  box-shadow: 0px 4px 5px rgba(0, 0, 0, 0.1);
}

.c4 {
  background: inherit;
  border: 0;
  font: inherit;
  text-align: left;
  cursor: pointer;
  position: relative;
  width: 100%;
  height: 60px;
  padding-left: calc(1em * 41 / (16 * 1.2));
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0.75em;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0.75em;
}

.c4 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c4 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c4 >span {
  font-size: 1em;
}

.c4::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: unset;
  width: 1em;
  background: green;
}

.c4:focus-visible {
  outline: 1px green solid;
}

.c5 {
  background: inherit;
  border: 0;
  font: inherit;
  text-align: left;
  cursor: pointer;
  position: relative;
  width: 100%;
  height: 60px;
  padding-left: calc(1em * 41 / (16 * 1.2));
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0.75em;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0.75em;
}

.c5 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c5 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c5 >span {
  font-size: 1em;
}

.c5::before {
  content: '';
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: unset;
  width: 1em;
  background: red;
}

.c5:focus-visible {
  outline: 1px red solid;
}

<div
  className="c0"
  onAnimationEnd={[Function]}
  variant="default"
>
  <div
    className="c1"
    onClick={[Function]}
    variant="default"
  >
    <svg
      fill="none"
      height="20"
      viewBox="0 0 22 25"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0.499998 0L10.5 -1.74845e-06C16.8513 -2.85895e-06 22 5.14872 22 11.5L22 13.4677C22 19.819 16.8513 24.9677 10.5 24.9677L0.500002 24.9677L0.499998 0Z"
        fill="#686867"
        opacity="0.8"
      />
      <path
        d="M7.0193 7.4903L11.9435 11.4297C12.6191 11.9701 12.6191 12.9976 11.9435 13.538L7.0193 17.4774"
        stroke="white"
        strokeLinecap="round"
        strokeWidth="0.9"
      />
    </svg>
  </div>
  <div
    className="c2"
  >
    test heading (undefined)
  </div>
  <div
    className="c3"
  >
    <button
      background="inherit"
      className="c4"
      flashColor="green"
      height="60px"
      id="id1"
      onClick={[Function]}
      orientation="left"
      width="100%"
    >
      <label
        htmlFor="id1"
      >
        test no sub
      </label>
    </button>
  </div>
  <div
    className="c3"
  >
    <button
      background="inherit"
      className="c5"
      flashColor="red"
      height="60px"
      id="id2"
      onClick={[Function]}
      orientation="left"
      width="100%"
    >
      <label
        htmlFor="id2"
      >
        test-tittle suite
      </label>
      <span>
        this is subtitle
      </span>
    </button>
  </div>
</div>
`;

exports[`SidePanel hyproof 1`] = `
.c3 {
  color: #4a4a4a;
  font-family: Roboto;
  margin-bottom: 100px;
  max-width: 80%;
  font-size: 30px;
  line-height: 45px;
}

.c2 {
  position: absolute;
  top: 10%;
  right: 0px;
  cursor: pointer;
  transform: scaleX(1) translateX(100%);
  transform-origin: 100%;
  transition: transform 0.7s ease-in-out;
}

.c4 {
  transition: all 0.3s ease-out;
}

.c4:hover {
  opacity: 0.7;
}

.c0 {
  position: fixed;
  box-sizing: border-box;
  width: 300px;
  height: 100lvh;
  left: 0px;
  top: 0px;
  display: flex;
  flex-direction: column;
  gap: 0.2em;
  padding: 5px 25px;
  transform: translateX(     -300px   );
  transition: transform 0.7s ease-in-out;
  box-shadow: 0px 4px 5px rgba(0, 0, 0, 0.1);
}

.c1 {
  box-shadow: none;
  background: rgba(255, 255, 255, 90%);
  backdrop-filter: blur(5px);
}

.c5 {
  display: flex;
  flex-direction: row;
  align-items: center;
  justify-content: flex-start;
  box-sizing: content-box;
  width: 90%;
  overflow: hidden;
  background: #d9d9d9;
  border: 2px solid color-mix(in lab, #d9d9d9 70%, black);
  border-radius: 60px;
  padding: 5px;
}

.c7 {
  background: inherit;
  border: 0;
  font: inherit;
  text-align: left;
  cursor: pointer;
  position: relative;
  width: 80%;
  padding-left: 10px;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0.75em;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0.75em;
}

.c7 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c7 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c7 >span {
  font-size: 1em;
}

.c7::before {
  content: '';
  position: absolute;
  display: none;
  top: 0;
  bottom: 0;
  left: 0;
  right: unset;
  width: 1em;
  background: green;
}

.c7:focus-visible {
  outline: 1px green solid;
}

.c9 {
  background: inherit;
  border: 0;
  font: inherit;
  text-align: left;
  cursor: pointer;
  position: relative;
  width: 80%;
  padding-left: 10px;
  overflow: hidden;
  display: flex;
  flex-direction: column;
  justify-content: center;
  gap: 0.5em;
  border-top-left-radius: 0.75em;
  border-top-right-radius: 0;
  border-bottom-right-radius: 0;
  border-bottom-left-radius: 0.75em;
}

.c9 >* {
  display: block;
  padding: 0;
  margin: 0;
  width: 90%;
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
  cursor: pointer;
}

.c9 >label {
  font-size: 1.2em;
  font-weight: bolder;
}

.c9 >span {
  font-size: 1em;
}

.c9::before {
  content: '';
  position: absolute;
  display: none;
  top: 0;
  bottom: 0;
  left: 0;
  right: unset;
  width: 1em;
  background: red;
}

.c9:focus-visible {
  outline: 1px red solid;
}

.c6 {
  box-sizing: border-box;
  display: grid;
  align-items: center;
  justify-content: center;
  width: 60px;
  font-size: calc(60px / 4);
  text-transform: uppercase;
  aspect-ratio: 1/1;
  border-radius: 50%;
  border: calc(60px * 0.065) solid white;
  background: green;
}

.c8 {
  box-sizing: border-box;
  display: grid;
  align-items: center;
  justify-content: center;
  width: 60px;
  font-size: calc(60px / 4);
  text-transform: uppercase;
  aspect-ratio: 1/1;
  border-radius: 50%;
  border: calc(60px * 0.065) solid white;
  background: red;
}

<div
  className="c0 c1"
  onAnimationEnd={[Function]}
  variant="hyproof"
>
  <div
    className="c2"
    onClick={[Function]}
    variant="hyproof"
  >
    <svg
      fill="none"
      height="20"
      viewBox="0 0 22 25"
      width="20"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M0.499998 0L10.5 -1.74845e-06C16.8513 -2.85895e-06 22 5.14872 22 11.5L22 13.4677C22 19.819 16.8513 24.9677 10.5 24.9677L0.500002 24.9677L0.499998 0Z"
        fill="#686867"
        opacity="0.8"
      />
      <path
        d="M7.0193 7.4903L11.9435 11.4297C12.6191 11.9701 12.6191 12.9976 11.9435 13.538L7.0193 17.4774"
        stroke="white"
        strokeLinecap="round"
        strokeWidth="0.9"
      />
    </svg>
  </div>
  <div
    className="c3"
  >
    test heading (undefined)
  </div>
  <div
    className="c4"
  >
    <button
      active={false}
      background="#d9d9d9"
      className="c5"
      onClick={[Function]}
      variant="hyproof"
    >
      <div
        bgColor="green"
        className="c6"
        fullName="unit"
        outlineColor="white"
        size="60px"
      >
        <p>
          u
        </p>
      </div>
      <button
        background="inherit"
        className="c7"
        flashColor="green"
        height="60px"
        id="id3"
        orientation="left"
        variant="hyproof"
        width="100%"
      >
        <label
          htmlFor="id3"
        >
          test no sub
        </label>
      </button>
    </button>
  </div>
  <div
    className="c4"
  >
    <button
      active={false}
      background="#d9d9d9"
      className="c5"
      onClick={[Function]}
      variant="hyproof"
    >
      <div
        bgColor="red"
        className="c8"
        fullName="test"
        outlineColor="white"
        size="60px"
      >
        <p>
          t
        </p>
      </div>
      <button
        background="inherit"
        className="c9"
        flashColor="red"
        height="60px"
        id="id4"
        orientation="left"
        variant="hyproof"
        width="100%"
      >
        <label
          htmlFor="id4"
        >
          test-tittle suite
        </label>
        <span>
          this is subtitle
        </span>
      </button>
    </button>
  </div>
</div>
`;
